crates/ironrdp-graphics/src/rdp6/rle.rs

```diff
@@ -152,13 +162,221 @@ pub fn decompress_8bpp_plane(
+            } else {
+                match count {
+                    3.. => {
+                        written += self
+                            .encode_segment(&raw, count, dst)
+                            .ok_or(RleEncodeError::BufferTooSmall)?;
+                        raw.clear();
+                    }
+                    2 => raw.extend_from_slice(&[last, last]),
+                    1 => raw.push(last),
+                    _ => {}
+                }
+
+                raw.push(byte);
+
+                (byte, 0)
+            }
+        }
+
+        let (last, mut count) = seq;
+        if count < 3 {
+            raw.extend(vec![last; count].into_iter());
+            count = 0;
+        }
+
+        written += self
+            .encode_segment(&raw, count, dst)
+            .ok_or(RleEncodeError::BufferTooSmall)?;
+
+        Ok(written)
+    }
+
+    fn encode_segment(&self, mut raw: &[u8], run: usize, dst: &mut WriteCursor<'_>) -> Option<usize> {
+        let mut extra_bytes = 0;
+
+        while raw.len() > 15 {
+            extra_bytes += self.encode_segment(&raw[0..15], 0, dst)?;
+            raw = &raw[15..];
+        }
+
+        let control = ((raw.len() as u8) << 4) + std::cmp::min(run, 15) as u8;
+
+        ensure_size!(dst: dst, size: raw.len() + 1);
+
+        dst.write_u8(control);
+        dst.write_slice(raw);
+
+        if run > 15 {
+            let last = raw.last().unwrap();
+            extra_bytes += self.encode_long_sequence(run - 15, *last, dst)?;
+        }
+
+        Some(1 + raw.len() + extra_bytes)
+    }
+
+    fn encode_long_sequence(&self, mut run: usize, last: u8, dst: &mut WriteCursor<'_>) -> Option<usize> {
+        let mut written = 0;
+
+        while run >= 16 {
+            ensure_size!(dst: dst, size: 1);
+
+            let current = std::cmp::min(run, MAX_DECODED_SEGMENT_SIZE) as u8;
+
+            let c_raw_bytes = std::cmp::min(current / 16, 2);
+            let n_run_length = current - c_raw_bytes * 16;
+
+            let control = (n_run_length << 4) + c_raw_bytes;
+            dst.write_u8(control);
+            written += 1;
+
+            run -= current as usize;
+        }
+
+        if run > 0 {
+            match run {
+                short @ 1..=3 => {
+                    written += self.encode_segment(&vec![last; short], 0, dst)?;
+                }
+                long => {
+                    written += self.encode_segment(&[last], long - 1, dst)?;
+                }
+            }
+        }
+
+        Some(written)
+    }
+}
+
+/// Performs compression of 8bpp color plane pixel stream into a buffer.
+/// Pixel iterator must have at least width * height items.
+/// Destination slice must have enough space for the compressed data.
+///
+/// Returns number of bytes written to the dst buffer.
+pub fn compress_8bpp_plane(
+    src: impl Iterator<Item = u8>,
+    dst: &mut WriteCursor<'_>,
+    width: impl Into<usize>,
+    height: impl Into<usize>,
```

Comment:
What about taking `usize` without generics here?

```suggestion
    width: usize,
    height: usize,
```

I’m not sure we want to hide this kind of conversion.

Reply:
Oh, I didn’t notice. Not a big deal, but if you could change `decompress_8bpp_plane` the same way, I would appreciate it.